NETWORK_PROTOCOL.md

```diff
@@ -1 +1,110 @@
+
+## The Reactive Streams core protocol
+
+The basic RS signalling is:
+
+    --> subscribe(publisher: String, subscriber: Id, initialDemand: Long = 0)
+    --> request(subscriber: Id, demand: Long)
+    --> cancel(subscriber: Id)
+    <-- onSubscribe(subscriber: Id, elementSize: varint = 0) // For elementSize != 0, see the next section
+    <-- onNext(subscriber: Id, element: bytes) 
+    <-- onComplete(subscriber: Id)
+    <-- onError(subscriber: Id, error: String)
+    
+The protocol is fully bidirectional; either party can act in the `-->` direction. The semantics for ordering and asynchronous delivery are the same as in the Reactive Streams specification.
+
+Unlike in RS, there is no separate Subscription object; the subscriber Id identifies the recipient in all messages going <-- this way. This id is generated by the subscriber and sent in the `subscribe` message.
+
+The publisher String needs to be parsed by the recipient; it is not described by this specification. [Could be added?]
+
+The field `onSubscribe.elementSize`, if nonzero, indicates the fixed size of the elements that will be published in this stream. In fixed-size mode, the `onNext.element` field is not length-prefixed. This saves space when the messages are very small, such as individual ints.
+
+After a subscription is closed, its Id can be reused, to prevent Ids from growing without limit. The subscriber MAY reuse an Id in a `subscribe` message after it has sent `cancel` or received `onComplete` or `onError` for that Id. If it does so, it MUST guarantee that the publisher will not receive messages meant for the previous subscription with that Id after it receives the second `subscribe` message.
+
+## Packed messaging
+
+In typical use, the most common messages by far are `onNext`. The overhead per message is typically 1 byte (message code) +  1-2 bytes (subscriber id) + 1-3 bytes (payload length) = 3-6 bytes total. When the message type is very small (e.g. an int), the overhead can be 100% or more.
+
+To reduce the overhead, the publisher can optionally declare that all stream elements will have a fixed size by setting the `onSubscribe.elementSize` field to a value greater than zero:
+
+    <-- onSubscribe(subscriber: Id, elementSize: varint)
+
+The publisher can then send not just `onNext` messages but also `onNextPacked` messages:
+
+    <-- onNextPacked(subscriber: Id, count: varint, messages: count * elementSize bytes)
+    
+Packing does not help if new data becomes available very frequently and must not be delayed before being sent. A typical example is a ticker source. It also can't be done if the client doesn't provide enough demand.
+
+## Split messaging
```

Comment:
The max size of a "fragment" will need to be defined. I would suggest no larger than 64K minus some overhead. That way it can be kept to a single 2 byte length.